### PR TITLE
COMMONS-500: Loop on es indexation if an error occurs

### PR DESCRIPTION
**Pull request by boubaker**
The current behavior seems not good for an indexation. (Stop the whole indexation operation on fail)
The new behavior will be:
- Continue indexing when fail + log the failed operation.
- Don't attempt to reindex the failed operation.

Another problem is fixed at the same time. In fact, the delete operation is treated in one single block and not in limited bulk size. So "checkBulkRequestSizeReachedLimitation" is added in "DELETE" block operation too.

### CODE REVIEW

File: commons-search/src/main/java/org/exoplatform/addons/es/index/impl/ElasticIndexingOperationProcessor.java [hunk 1]
```diff
@@ -191,16 +191,33 @@ private void processCUD(Map<OperationType, Map<String, List<IndexingOperation>>>
...
+          continue;
+        }
+        Iterator<IndexingOperation> deleteIndexingOperationsIterator = deleteIndexingOperationsList.iterator();
+        while (deleteIndexingOperationsIterator.hasNext()) {
+          IndexingOperation deleteIndexQueue = deleteIndexingOperationsIterator.next();
+          try {
+            bulkRequest += elasticContentRequestBuilder.getDeleteDocumentRequestContent((ElasticIndexingServiceConnector) getConnectors().get(deleteIndexQueue.getEntityType()),
+                                                                                        deleteIndexQueue.getEntityId());
+          } catch(Exception e) {
+            LOG.warn("Error while *deleting* index of entity, type = " + entityType + ", id =" + (deleteIndexQueue == null ? null : deleteIndexQueue.getEntityId()) + ", cause:", e);
```
Comment (thomasdelhomenie) [inline commons-search/src/main/java/org/exoplatform/addons/es/index/impl/ElasticIndexingOperationProcessor.java]: The message is misleading IMO, it should be something like "Error while *deleting* index entry of entity ..." instead.

Comment (boubaker) [inline commons-search/src/main/java/org/exoplatform/addons/es/index/impl/ElasticIndexingOperationProcessor.java]: Do you need to add "entry" word ?

Comment (thomasdelhomenie) [inline commons-search/src/main/java/org/exoplatform/addons/es/index/impl/ElasticIndexingOperationProcessor.java]: Your message states that you will delete an index, not just a document in the index.

**Comment by boubaker** [comment]
> Can we add an unit test to check that the process continue on failure, and that the log is done ?

FYI, it will not continue on all failures.
It will continue just when the bulkRequest variable update failed (like we had on community). When the ES server is down, for example, it will retry until the server is up. Same thing when the ES cannot index a document.

**Comment by thomasdelhomenie** [comment]
So if I understand well :
* when the bulkRequest variable update fails, the document will not be indexed, a log is written, and the operation will be deleted from the queue table so it will not be processed again
* when the ES instance cannot be reached or any other error is thrown during the indexation request, the process is stopped and the operations will are not deleted in the operations table, so it will be played again at the next job execution (so every 5 seconds by default)

Right ?

**Comment by boubaker** [comment]
> So if I understand well ...

Yes.

**Comment by thomasdelhomenie** [comment]
In the initial tech spec, we stated :

> In case of error, the entityID+entityType will be logged in a “error queue” to allow a manual reprocessing of the indexing operation. However, in a first version of the implementation, the error will only be logged with ERROR level in the log file of platform and an event will be inserted in the exo-es-search audit trail.

But in the case of an error on the indexation request (so for example when ES is not available), it is probably better to keep the indexation operation in the queue (as done in this PR) and retry at the next iteration.
We could also use a timing algorithm such as exponential backoff in a future improvement.

@vsellier WDYT ?

**Comment by boubaker** [comment]
The indexing operation can fails all the time because of previous blocking indexing operations (that wasn't deleted) and it will not index data anymore.
In other words, this can lead to those two cases:
*  the queue may be filled with blocking operations (operations in error == BATCH_NUMBER) and can no longer index
* each indexing iteration (a bulk), we will reattempt the failed operations and we will have a lot of error logs

**Comment by vsellier** [comment]
IMO, we should follow these rules :

-|-
-----|-----
ES not reachable | retry, the adaptable delay is a good thing, but we should also have a status to disable the indexation and the search (for ES maintenance or unexpected behavior)
Error during the message preparation (connector) | discard if it's not a runtime problem (database not available, ...)
ES error during the indexation | log and discard as today

A good improvement for later could be to keep in database the discarded operations with a specific status and improve the management pages to allow the drop or the retry manually.
I don't know if it's included in 
>In case of error, the entityID+entityType will be logged in a “error queue” to allow a manual reprocessing of the indexing operation.
;)

**Comment by boubaker** [comment]
Thank you @vsellier for the feedback.
So the solution will be:
To implement point 1 (ES not reachable)
* Add a Boolean flag in ES Client to describe ES availability.
* Update Social & Wiki Service layer to test if the ES is available or not (with the introduction of CustomExcepion on read/write operations so that we can switch the Service layers to use Database)
* In case of ES unavailability:
** trace a log error 
** switch the flag of ES availability to false
** Add a scheduled thread that tests on ES availability (each 30 seconds for example) & treat the indexation operations from queue & then switch back the flag to true (so that Social & Wiki layers use it again).
For point 2 & 3 (ES not reachable):
* in case of an error, keep the indexation operation in queue, but add a flag "Error" for all operation existing in the bulk & continue the indexation
* Make the indexation Job ignore all previous operations from queue that are flagged as error
* Add a new job that will have a different periodicity (2 hours) to try again the indexation operation one by one and not by bulk.

@vsellier , @thomasdelhomenie WDYT ?

**Comment by thomasdelhomenie** [comment]
> Add a scheduled thread that tests on ES availability (each 30 seconds for example) & treat the indexation operations from queue & then switch back the flag to true (so that Social & Wiki layers use it again).

I don't think we need a new scheduled job here. We already have the indexing job which can be improved with an adaptable delay. WDYT ?

> For point 2 & 3 (Error during the message preparation (connector) & ES error during the indexation )

I think we should open new issues to treat the others improvements, but maybe that was your intention, bot sure :smile:

**Comment by boubaker** [comment]
> I don't think we need a new scheduled job here. We already have the indexing job which can be improved with an adaptable delay. WDYT ?

I didn't mean a "scheduled job" managed by Quartz, but a scheduled Thread (ScheduledExecutor) that is executed independently from Indexation/Reading ES. But I agree that it can be managed with Indexing Job. Thus, I don't think that the adaptable time is necessary here since we need to periodically (I prefer to use fixed period because it's not expensive in time of resources) check for availability of ES without having to increase the periodicity of this check.

> I think we should open new issues to treat the others improvements, but maybe that was your intention, bot sure :smile:

Nope, it wasn't my intention. But if you want, yes, we can manage it in different issue.

**Comment by thomasdelhomenie** [comment]
As discussed, this PR (and the related issue COMMONS-524) must implement the following points :
* in case of an error, keep the indexation operation in queue, but add a flag "Error" for all operation existing in the bulk & continue the indexation
* Make the indexation Job ignore all previous operations from queue that are flagged as error
* Add a new job that will have a different periodicity (2 hours) to try again the indexation operation one by one and not by bulk.

Others items discussed in this PR will be implemented in COMMONS-500.

**Comment by thomasdelhomenie** [comment]
As discussed, the scope of COMMONS-500 is to make sure the indexation operation does not fail and loop. All others points (like those listed in https://github.com/exodev/commons/pull/62#issuecomment-274496027) will be treated in COMMONS-524.

**Comment by thomasdelhomenie** [comment]
Merged.

**Pull request closed without merging**